base/abstractarray.jl

```diff
@@ -1188,7 +1188,9 @@ typed_hcat(::Type{T}, X::Number...) where {T} = hvcat_fill(Matrix{T}(undef, 1,le
 vcat(V::AbstractVector...) = typed_vcat(promote_eltype(V...), V...)
 vcat(V::AbstractVector{T}...) where {T} = typed_vcat(T, V...)
 
-function typed_vcat(::Type{T}, V::AbstractVector...) where T
+AbstractVecOrTuple{T} = Union{AbstractVector{<:T}, Tuple{Vararg{T}}}
```

Comment:
Whoa, crazy — this works?  You can make a type covariant through a union?

In any case, I think I'd prefer doing this the other way around — I saw `A::AbstractVecOrTuple{AbstractVecOrMat}` and assumed it was an invariance bug.  That is:

```julia
AbstractVecOrTuple{T} = Union{AbstractVector{T}, Tuple{Vararg{T}}}
# and
_typed_hcat(::Type{T}, A::AbstractVecOrTuple{<:AbstractVecOrMat}) where T
```

Reply:
Good point. Unfortunately, I get a bootstrap error if I add `<:` in the method signatures instead (whether I also change `<:T` to `T` in the type alias or not). That's weird since I cannot reproduce at the REPL. Maybe that's a dispatch bug.

```
REPL  ────────────────── 17.461338 seconds
error during bootstrap:
LoadError("sysimg.jl", 213, LoadError("/home/milan/Dev/julia/usr/share/julia/stdlib/v0.7/Pkg3/src/Pkg3.jl", 32, LoadError("/home/milan/Dev/julia/usr/share/julia/stdlib/v0.7/Pkg3/src/Types.jl", 45, MethodError(Base._typed_vcat, (UInt8, (UInt8[0xc8, 0x30, 0xd4, 0x4f, 0xc0, 0x00, 0xb4, 0x80, 0xd1, 0x11, 0xad, 0x9d, 0x10, 0xb8, 0xa7, 0x6b], UInt8[0x6a, 0x75, 0x6c, 0x69, 0x61, 0x6c, 0x61, 0x6e, 0x67, 0x2e, 0x6f, 0x72, 0x67])), 0x00000000000067bb))))
rec_backtrace at /home/milan/Dev/julia/src/stackwalk.c:94
record_backtrace at /home/milan/Dev/julia/src/task.c:246
jl_throw at /home/milan/Dev/julia/src/task.c:577
jl_method_error_bare at /home/milan/Dev/julia/src/gf.c:1608
jl_method_error at /home/milan/Dev/julia/src/gf.c:1626
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:2117
typed_vcat at ./abstractarray.jl:1271
vcat at ./abstractarray.jl:1189 [inlined]
uuid5 at /home/milan/Dev/julia/usr/share/julia/stdlib/v0.7/Pkg3/src/Types.jl:36
```

Reply:
My hunch would be an ambiguity error?

Reply:
Yeah but as I said I can't reproduce when pasting the definitions directly at the REPL.

Reply:
Filed as https://github.com/JuliaLang/julia/issues/27224.

Reply:
Should

```
AbstractVecOrTuple{T} = Union{AbstractVector{<:T}, Tuple{Vararg{T}}}
```

be declared `const`  ?

Reply:
If there is a parameter it doesn't matter. Personally I always put const for consistent.